Standardize tests for abbr tag, improve N.B., P.S., am/pm, eoc checks

vr8hub opened the pull request:
If you want to save this until after the next release, that's no problem, but I wanted to go ahead and get it in so I don't forget.

Improvements:
1. The N.B. search expression ended with `\.\b`. Since the period was not optional, that turned the `\b` into the wrong test; since the period is a word boundary, it essentially forced the following character to NOT be a word boundary. Which isn't what we want. So I changed it to a `\B`.
2. I added two additional am/pm tests looking for a number prior, and either upper or lower case am/pm with or without periods. This lets us catch a number of additional cases (see the first section in the updated attached test file), and I don't believe will give us any more false positives (or maybe very, very few). I created two since python doesn't support \L to lowercase a replacement string; I could have created a function and called it, but I thought a separate test was cleaner and clearer.
3. The page regex was matching just the p. in "p. m.", so I moved it after the am/pm checks and put a comment why it was there and that it shouldn't be moved prior.
4. The tests for an existing abbr tag were inconsistent; some were just `<abbr>`, some tried to look for specific things in the tag, some looked for anything. I standardized them all to look for anything; we don't really care what's in the `<abbr>` tag; if the tag is there, we don't want to process the abbreviation again.
5. The P.S. regex was matching U.P.S., so I modified the regex to not match on a preceding period or a following non-word boundary (again, because we're looking for an unconditional trailing period).
6. Since we only have the "eoc" class left for abbreviations, I eliminated worrying about an existing class. The possibility of having a styling class on an abbreviation tag is so remote I didn't think the cost of worrying about it was worth it.
7. I consolidated the etc. checks into a single regex.
8. I added an additional "any abbreviation ending in a period" check for eoc that is essentially the same as the etc. check, but without the following uppercase letter portion, per our previous discussion.

I updated the attached test file to include tests for all of the above, and this version passes all the tests. I also ran it on four of our largest books to make sure it wasn't generating a bunch of false positives: Pepys, O. Henry, Les Mis, War & Peace. It did catch a (very) few things, some of which would have been caught before, and it generated a handful of false positives, all of which are generated by the existing code. So, no worse than existing, and a little better on already semanticated text.

I'll work on updating the semanticate test next to use my test file, but it might be later next week.

[test-semanticate.txt](https://github.com/standardebooks/tools/files/7071318/test-semanticate.txt)

acabal commented on se/formatting.py:
```diff
@@ -50,78 +50,76 @@ def semanticate(xhtml: str) -> str:
+	xhtml = regex.sub(r"""(?<!\<abbr[^\>]*?\>)\bN\.?B\.\B""", r"""<abbr epub:type="z3998:initialism">N.B.</abbr>""", xhtml)
+	xhtml = regex.sub(r"""(?<!\<abbr[^\>]*?\>)Ph\.?\s*D\.?""", r"""<abbr epub:type="z3998:name-title">Ph. D.</abbr>""", xhtml)
+	xhtml = regex.sub(r"""(?<!\<abbr[^\>]*?\>)(?:IOU(?:\.|\b)|I\.O\.U\.)""", r"""<abbr epub:type="z3998:initialism">I.O.U.</abbr>""", xhtml)
+	xhtml = regex.sub(r"""(?<!\<abbr[^\>]*?\>)T\.?V\.?\b""", r"""<abbr epub:type="z3998:initialism">TV</abbr>""", xhtml)
```
This is the same pattern as N.B. but the regex differs. Should we make them the same?

vr8hub replied:
The TV? Yes, almost certainly; I'll see if I can add a couple of cases and test that tomorrow, and look around for any others.
And I forgot something in my improvements list: The IOU regex before was creating an extra period in some cases, e.g. `<p>This is an IOU.</p>` was turned into `<p>This is an <abbr epub:type="z3998:initialism">I.O.U.</abbr>.</p> I changed it so that it left only a single period in all cases (and only allow IOU or I.O.U.). The eoc check will then pick it up (it wasn't before because of the extra period).

vr8hub replied:
Grr, I forgot to tag the "turned into." That should be "…turned into `<p>This is an <abbr epub:type="z3998:initialism">I.O.U.</abbr>.</p>`".

vr8hub commented:
I've pushed some additional changes. TV was more like AD/BC than N.B. Since we remove any periods from TV, so we have to make sure we keep a trailing period if it's an eoc. But in the process of testing, and finding and fixing some false positives, I also changed the lookahead on everything. It now looks for a preceding period, `\B`, or the abbr tag. IOW, what precedes an abbreviation should be a word-boundary that isn't a period (in case it's part of a longer abbreviation), and can't be an existing abbr tag.

I removed any `\b` that were at the front of existing ones (since the above does the same thing but more accurately), and the `\s` at the front of a few (TV, for one), since that kept us from catching things we want to catch, e.g. `<p>TV is a vast wasteland.</p>`. And I added a few tests to confirm the false-positives were corrected, e.g. `A.B.C.` to confirm `B.C.` wasn't caught, etc., and everything passes in it. (There's a false-positive in the "should not get tag" section, but that's from roman numerals, not abbreviations.)

I tested the same big four again, and again they look good. I also made a couple of copies of Les Mis, took the first copy back to just after the semanticate, the second copy to just before and then ran the new semanticate, then compared every file. There were only a handful of differences, and they were all either valid or false positives from other sections of the code that were unchanged, e.g. roman numerals (which I have plans for, but after I get the semanticate test file integrated).

In the process I found two books that have se:era abbreviations that still have the periods in them: Pepys and Armagaddon 2419 AD. Since the latter is in the title, it might get messy, so I'm going to leave those to you if you don't mind. They're not time-sensitive, obviously. :)
[test-semanticate.txt](https://github.com/standardebooks/tools/files/7073272/test-semanticate.txt)

vr8hub commented:
I replaced the existing semanticate test file with mine, and put the results of semanticate on that file in the out directory. I ran pytest locally and everything worked, so I _think_ that's all there is. Let me know if not.

acabal commented:
Looks good, great work, thanks!
